File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.py
import heapq
import sys

input = sys.stdin.readline

dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]


def dijkstra(x):
    global n
    distance = [1e9] * (n * n)
    distance[x] = arr[x // n][x % n]

    q = []
    heapq.heappush(q, [arr[x // n][x % n], x])

    while q:
        dist, now = heapq.heappop(q)

        if distance[now] < dist:
            continue

        for i in li[now]:
            cost = dist + i[1]

            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q, [cost, i[0]])

    return distance

idx = 1
while True:
    n = int(input())
    if n == 0:
        break;

    li = [[] for _ in range(n * n)]  # 2차원배열이니까 n*n개
    arr = [list(map(int, input().split())) for _ in range(n)]

    for i in range(n * n):
        x = i // n;
        y = i % n;
        for dir in range(4):
            nx = x + dx[dir];
            ny = y + dy[dir];

            # 범위 벗어나면 종료
            if nx < 0 or ny < 0 or nx >= n or ny >= n:
                continue;
            # 시작점에 -> [도착점, 가중치]
            li[i].append([nx * n + ny, arr[nx][ny]])

    res = dijkstra(0) # 0, 0이 시작점
    print(f"Problem {idx}: {res[-1]}")
    idx += 1

File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n;
    static ArrayList<Node>[] li;
    static int[][] arr;

    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {-1, 1, 0, 0};

    static class Node implements Comparable<Node> {
        int vertex;
        int distance;

        public Node(int vertex, int distance) {
            this.vertex = vertex;
            this.distance = distance;
        }

        @Override
        public int compareTo(Node other) {
            return Integer.compare(this.distance, other.distance);
        }
    }

    static boolean setInputs() throws IOException {
        n = Integer.parseInt(br.readLine());
        if (n == 0) {
            return false;
        }

        li = new ArrayList[n * n];
        for (int i = 0; i < n * n; i++) {
            li[i] = new ArrayList<>();
        }

        arr = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0; i < n * n; i++) {
            int x = i / n;
            int y = i % n;
            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];

                if (nx < 0 || ny < 0 || nx >= n || ny >= n) {
                    continue;
                }

                li[i].add(new Node(nx * n + ny, arr[nx][ny]));
            }
        }
        return true;
    }

    public static int[] dijkstra(int start) {
        int[] distance = new int[n * n];
        for (int i = 0; i < n * n; i++) {
            distance[i] = Integer.MAX_VALUE;
        }
        distance[start] = arr[start / n][start % n];

        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(start, arr[start / n][start % n]));

        while (!pq.isEmpty()) {
            Node node = pq.poll();
            int dist = node.distance;
            int now = node.vertex;

            if (distance[now] < dist) {
                continue;
            }

            for (Node next : li[now]) {
                int nextVertex = next.vertex;
                int cost = dist + next.distance;

                if (cost < distance[nextVertex]) {
                    distance[nextVertex] = cost;
                    pq.add(new Node(nextVertex, cost));
                }
            }
        }
        return distance;
    }

    public static void main(String[] args) throws IOException {
        int idx = 1;
        while (setInputs()) {
            int[] res = dijkstra(0);
            sb.append("Problem ").append(idx).append(": ").append(res[res.length - 1]).append("\n");
            idx++;
        }

        System.out.print(sb);
    }
}
